File: cses/BFS_message_route.py
#* ************************************************************************** *#
#*                                                                            *#
#*                            \\             #`#``                            *#
#*                            ~\o o_       0 0\                               *#
#*                            # \__)      (u  ); _  _                         *#
#*                     # \# \#  #           \  \# \# \                        *#
#*                    #(   . . )            (         )\                      *#
#*                   #  \_____#              \_______#  \                     *#
#*                       []  []               [[] [[]    *.                   *#
#*                       []] []]              [[] [[]                         *#
#*                                                                            *#
#* ****************************************************************** nuo *** *#

import sys

n, m = [int(x) for x in input().split()]

E = [[] for _ in range(n)]

for _ in range(m):
    a, b = [int(x) for x in input().split()]
    a -= 1
    b -= 1
    E[a].append(b)
    E[b].append(a)

Parent = [None for _ in range(n)]
Seen = [False for _ in range(n)]
Seen[0] = True
Q = [0]

while Q:
    #print(Parent)
    x = Q.pop(0)
    if x == n - 1:
        route = [x]
        while Parent[x] is not None:
            x = Parent[x]
            route.append(x)
        route = [str(x + 1) for x in reversed(route)]
        print(len(route))
        print(' '.join(route))
        sys.exit(0)

    for i in E[x]:
        if Seen[i]:
            continue
        Seen[i] = True
        Parent[i] = x
        Q.append(i)

print('IMPOSSIBLE')

File: cses/BFS_message_route.cpp
/* ************************************************************************** */
/*                                                                            */
/*                            \\             /`/``                            */
/*                            ~\o o_       0 0\                               */
/*                            / \__)      (u  ); _  _                         */
/*                     / \/ \/  /           \  \/ \/ \                        */
/*                    /(   . . )            (         )\                      */
/*                   /  \_____/              \_______/  \                     */
/*                       []  []               [[] [[]    *.                   */
/*                       []] []]              [[] [[]                         */
/*                                                                            */
/* ****************************************************************** nuo *** */

#include "iostream"
#include "vector"
//#include "stack" // stack push: to the top
#include "queue" // queue push: to the end

using   namespace std;

int     main(void)
{
        queue<int>  Q;
        string      s;
        int         n, m, x, y, i;

        cin >> n >> m;

        vector<vector<int>> E(n);
        vector<int>         parent(n, 0);
        vector<int>         seen(n, false);
        vector<int>         path;

        seen[0] = true;
        Q.push(0);

        i = 0;
        while (i++ < m)
        {
            cin >> x >> y;
            x -= 1;
            y -= 1;
            E[x].push_back(y);
            E[y].push_back(x);
        }

        while (!Q.empty())
        {
            x = Q.front();
            Q.pop();
            if (x == n - 1)
            {
                path.push_back(x);
                while (parent[x])
                {
                    x = parent[x];
                    path.push_back(x);
                }
                path.push_back(0);
                cout << path.size() << endl;
                i = (int) path.size();
                while (--i > -1)    cout << path[i] + 1 << ' ';
                cout << endl;
                return (0);
            }
            i = -1;
            while (++i < (int) E[x].size())
            {
                if (seen[E[x][i]])  continue;
                seen[E[x][i]] = true;
                parent[E[x][i]] = x;
                Q.push(E[x][i]);
            }
        }
        cout << "IMPOSSIBLE" << endl;
        return (0);
}
